Add "Friends" tab to social browser

Aergwyn opened the pull request:
As said in title this should lay the groundwork to be able to see your friends in the social browser.

Aergwyn committed:
initial commit to allow filtering / selecting different tabs / etc

still wip tho
bleh


bleh

Aergwyn committed:
fix selecting different tab causing filter to change by itself

and minor cleanup

Aergwyn committed:
center user panels

Aergwyn committed:
query friends endpoint to fetch friendlist

Aergwyn committed:
Updated submodule osu-resources

Aergwyn committed:
Updated submodule osu-framework

Aergwyn committed:
Merge branch 'master' into expand-social-tab

Aergwyn commented:
I guess other functionality for social would be more appropriate in a separate PR (and after I asked a few things).

peppy commented on osu.Game/Overlays/Social/Header.cs:
```diff
@@ -55,8 +55,8 @@ public enum SocialTab
         OnlinePlayers,
-        //[Description("Online Friends")]
-        //OnlineFriends,
+        [Description("Online Friends")]
```
Let's remove the "Online" part from these tabs for now, since it's not actually checking online status.

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -9,19 +9,23 @@
 
 namespace osu.Game.Overlays
 {
-    public class SocialOverlay : SearchableListOverlay<SocialTab, SocialSortCriteria, SortDirection>, IOnlineComponent
```
what was the reasoning for making this not an `IOnlineComponent`?

Aergwyn replied:
Because I removed the `APIStateChanged` it was no longer necessary. I didn't want a repeat from the leaderboard with endless reloads if the API fails. That should only be temporarily though.

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,62 +57,149 @@ public SocialOverlay()
         }
 
-        public void APIStateChanged(APIAccess api, APIState state)
+        private GetUsersRequest getUsersRequest;
```
i don't see any reason to store both of these requests separately – it just increases complexity. we should just store the currently running request for the currently displayed tab.

Aergwyn replied:
I tried with having just an `APIRequest` but as only `APIRequest<T>` has the Success event it won't work. Also both Requests return data in a different structure.
I could try merging those two request classes and just have the type fetched as an enum just like the other criterias would be once that comes. `new GetUsersRequest(UserType.Friends)`
UserType could include `Friends, Other, Team` , etc.
Does that sound better?

peppy reviewed (changes requested):
as mentioned

Aergwyn committed:
changed titles from social tabs

Aergwyn committed:
simplify socialoverlay user request structure

Aergwyn committed:
disable searching/filtering with textbox

it doesn't do anything except from creating unnecessary requests for now

Aergwyn committed:
readd IOnlineComponent to be able to react to APIState changes

Aergwyn committed:
remove unnecessary update

Aergwyn committed:
Merge branch 'master' into expand-social-tab

Aergwyn committed:
Updated submodule osu-framework

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,57 +57,146 @@ public SocialOverlay()
+
+        private readonly Bindable<string> currentQuery = new Bindable<string>();
+
+        private ScheduledDelegate queryChangedDebounce;
```
probably remove this as it isn't used currently

Aergwyn replied:
remove, as in commenting or actually (temporary) deleting?

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,57 +57,146 @@ public SocialOverlay()
+            switch (Header.Tabs.Current.Value)
+            {
+                case SocialTab.OnlinePlayers:
+                    getUsersRequest = new GetUsersRequest(); // TODO filter???
```
something like this will avoid casting:

```
var request = new GetUsersRequest();
request.Success += ...
api.Queue(getUsersRequest = request);
```

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,57 +57,146 @@ public SocialOverlay()
 
+        private void finishRequest(IEnumerable<User> newUsers)
+        {
+            Task.Run(() =>
```
what's the point of this `Task.Run`?

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,57 +57,146 @@ public SocialOverlay()
             loading.Show();
         }
 
+        private void finishRequest(IEnumerable<User> newUsers)
```
should probably be called `UpdateUsers` (or just move all this logic to the `Users` setter?)

Aergwyn committed:
change unnecessary cast

Aergwyn committed:
remove unnecessary task + rename method

Aergwyn committed:
make loading indicator more visible by deleting panels preemptively

Aergwyn committed:
remove unused code

Aergwyn committed:
added new tabs and sort criteria

Aergwyn committed:
fix SortDirection default to descending when retrieved data is ascending

Aergwyn committed:
bring social tab+filter behaviour closer to direct

Aergwyn committed:
enabled Location as filter tab

forgot it QQ

Aergwyn commented:
I guess panel animations are the last thing really missing for this PR.
Filtering and a bit of design for the panels themselves could be done in another one.

Aergwyn committed:
fix covers not showing in user panels

Aergwyn committed:
added hover effects to panels in social

at least partially QQ

Aergwyn committed:
Updated submodule osu-framework

Aergwyn committed:
Merge branch 'master' into expand-social-tab

Aergwyn commented:
Most likely two things left. Two things I'd like to have some assistance/advice with:
1) Move effect on hover. Exposing the content of `UserPanel` without breaking it somehow seems harder than it sounds
2) New panels are added too fast to the fillflow (before old ones are completely gone) which doesn't look nice. I haven't yet found a solution to this. Just simply delaying adding the new ones

Aergwyn committed:
fix new panels appearing too soon

Aergwyn committed:
update submodules

Aergwyn committed:
Merge branch 'master' into expand-social-tab

Aergwyn committed:
update licence headers

Aergwyn committed:
fix hover effects on social panels

Aergwyn commented:
Alright. It's ready for another review.

peppy committed:
Merge branch 'master' into expand-social-tab

Aergwyn committed:
tidying up

Aergwyn committed:
remove unnecessary Schedules

one was bugging out when rapidly switching display styles and the other was... unnecessary

Aergwyn commented:
I really can't get behind what to do to fix [this](https://streamable.com/54901).

peppy committed:
Merge branch 'master' into expand-social-tab

peppy commented:
The issue in your video is due to new elements fading in before the old ones complete their fade out (as long as they have an alpha above zero they will be considered for part of the flow).

peppy commented on osu.Game/Overlays/Social/Header.cs:
```diff
@@ -53,12 +54,12 @@ private void load(OsuColour colours)
-        //[Description("Online Team Members")]
-        //OnlineTeamMembers,
+        [Description("All Players")]
+        AllPlayers = SocialSortCriteria.Rank,
```
not sure about these `= ...` pieces here. should probably not be a thing.

Aergwyn replied:
Probably not needed anymore. Agreed.

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,65 +51,174 @@ public SocialOverlay()
+
+        private void updateSearch()
+        {
+            //queryChangedDebounce?.Cancel();
```
are we gonna add this back before merging? debounce should be present at least for search text updates.

peppy commented on osu.Game/Overlays/SocialOverlay.cs:
```diff
@@ -62,65 +51,174 @@ public SocialOverlay()
+            });
+        }
+
+        private void clearPanels()
```
To fix the animation issue you were having, you probably want to return `FadeOut`'s `TransformSequence` and attach the `FadeIn` of the next panels inside a `.Then.Schedule((() => {  })` or similar.

Aergwyn replied:
Nope. I can get a sequence but the content of `Schedule` after a `Then` is never called.
I thought it's because the old panels are Expiring and even set to null but that doesn't seem to affect anything.
Result is that I still can't get it to work.

peppy commented on osu.Game/Users/UpdateableAvatar.cs:
```diff
@@ -44,7 +44,7 @@ private void updateAvatar()
                 {
                     RelativeSizeAxes = Axes.Both,
-                    OnLoadComplete = d => d.FadeInFromZero(200),
+                    OnLoadComplete = d => d.FadeInFromZero(400, Easing.Out),
```
Considering this affects every avatar everywhere, 400 is probably a bit slow. Try 300 with `OutQuint`?

Aergwyn committed:
enable query

change Avatar animation too
add forgotten usings

peppy committed:
Merge branch 'master' into expand-social-tab

peppy committed:
Remove panel fade out for now

Should be implemented at one level above using a dedicated container.
